### PR TITLE
Timeline feature does not respect security constraint set by "add-group-repo" kie-cli command"

### PR DESCRIPTION
**Pull request by ederign**
Problem
------------
kie-config-cli can restrict access to the git repository/organization unit only for user with specific group. 

However, when a user - not a member of this group - displays the Home -> Timeline feature - he is still able to open & edit & save assets which are located in a protected git repository.

Solution
------------
In order to address this issue, I've to create a new plugabble security mechanism on kie-uberfire-extensions.
That mechanism includes a [SocialSecurityConstraintsManager](https://github.com/ederign/kie-uberfire-extensions/blob/ceb4e3a49f90c9f56e8c7818d43d85cc1bccb97c/kie-uberfire-social-activities/kie-uberfire-social-activities-backend/src/main/java/org/kie/uberfire/social/activities/security/SocialSecurityConstraintsManager.java)  that look for cdi implementations of [SocialSecurityConstraint](https://github.com/ederign/kie-uberfire-extensions/blob/ceb4e3a49f90c9f56e8c7818d43d85cc1bccb97c/kie-uberfire-social-activities/kie-uberfire-social-activities-api/src/main/java/org/kie/uberfire/social/activities/service/SocialSecurityConstraint.java) and apply this constraints to a list of social events.

[SocialSecurityConstraint](https://github.com/ederign/kie-uberfire-extensions/blob/ceb4e3a49f90c9f56e8c7818d43d85cc1bccb97c/kie-uberfire-social-activities/kie-uberfire-social-activities-api/src/main/java/org/kie/uberfire/social/activities/service/SocialSecurityConstraint.java)  has two methods, hasRestrictions that see if there is some restrictions for a given event, and init methods that is called before hasRestrictions in order to build some cache (if necessary).

Each [SocialTimelineCachePersistence](https://github.com/ederign/kie-uberfire-extensions/blob/ceb4e3a49f90c9f56e8c7818d43d85cc1bccb97c/kie-uberfire-social-activities/kie-uberfire-social-activities-backend/src/main/java/org/kie/uberfire/social/activities/persistence/SocialTimelineCachePersistence.java) that returns a timeline instead of returning events are now calling applySocialSecurityConstraints method that delegates to [SocialSecurityConstraintsManager](https://github.com/ederign/kie-uberfire-extensions/blob/ceb4e3a49f90c9f56e8c7818d43d85cc1bccb97c/kie-uberfire-social-activities/kie-uberfire-social-activities-backend/src/main/java/org/kie/uberfire/social/activities/security/SocialSecurityConstraintsManager.java).

The implementation of [SocialSecurityConstraint](https://github.com/ederign/kie-uberfire-extensions/blob/ceb4e3a49f90c9f56e8c7818d43d85cc1bccb97c/kie-uberfire-social-activities/kie-uberfire-social-activities-api/src/main/java/org/kie/uberfire/social/activities/service/SocialSecurityConstraint.java)  are located on kie-wb-common home page:

- [SocialEventOUConstraint](https://github.com/ederign/kie-wb-common/blob/8f071c1b9da2e42138456c02fad11b748be7d313/kie-wb-common-screens/kie-wb-common-social-home-page/kie-wb-common-social-home-page-backend/src/main/java/org/kie/workbench/common/screens/social/hp/security/SocialEventOUConstraint.java) that secures a event againts OU restrictions;
- [SocialEventRepositoryConstraint](https://github.com/ederign/kie-wb-common/blob/8f071c1b9da2e42138456c02fad11b748be7d313/kie-wb-common-screens/kie-wb-common-social-home-page/kie-wb-common-social-home-page-backend/src/main/java/org/kie/workbench/common/screens/social/hp/security/SocialEventRepositoryConstraint.java) that secures a project/asset event agains repositories permissions.

I also found and fix another feature leak issue related to link creations on Social Timelines. If a user has a permission to create a asset sometime, had created a asset and then lost the authoring permission, the links are still created and our user can assets AuthoringPerspective via social link. That was fixed in [DefaultSocialLinkCommandGenerator](https://github.com/ederign/kie-wb-common/blob/8f071c1b9da2e42138456c02fad11b748be7d313/kie-wb-common-screens/kie-wb-common-social-home-page/kie-wb-common-social-home-page-client/src/main/java/org/kie/workbench/common/screens/social/hp/client/homepage/DefaultSocialLinkCommandGenerator.java).

Child PR
------------
https://github.com/droolsjbpm/kie-wb-common/pull/132

### CODE REVIEW

**Comment by manstis** [comment]
No licenses!!!

I'll review more fully tomorrow.

**Comment by manstis** [comment]
I'm not keen on your use of constructors with a bazillion nulls; e.g. super(null, null, null, null....) but that's no reason to object to a PR. 

Other than licenses this gets a +1. Please add licenses and I'll ok this for merge.

**Comment by ederign** [comment]
Agree. The nulls that u are talking are the ones on test fakes right?

Em quarta-feira, 11 de novembro de 2015, Michael Anstis <
notifications@github.com> escreveu:

> I'm not keen on your use of constructors with a bazillion nulls; e.g.
> super(null, null, null, null....) but that's no reason to object to a PR.
>
> Other than licenses this gets a +1. Please add licenses and I'll ok this
> for merge.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/droolsjbpm/kie-uberfire-extensions/pull/9#issuecomment-155928920>
> .
>


-- 
Obrigado
_____________
Eder Ignatowicz
ignatowicz@gmail.com

**Comment by manstis** [comment]
Yes

**Comment by Rikkola** [comment]
I'll give notice on the best PR description of this week. Well done.

**Comment by ederign** [comment]
New files now have licenses. Old ones will be covered by https://github.com/droolsjbpm/kie-uberfire-extensions/pull/8

File: kie-uberfire-social-activities/kie-uberfire-social-activities-backend/src/test/java/org/kie/uberfire/social/activities/persistence/SocialTimelineCachePersistenceTest.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
+package org.kie.uberfire.social.activities.persistence;
```
Comment (manstis) [inline kie-uberfire-social-activities/kie-uberfire-social-activities-backend/src/test/java/org/kie/uberfire/social/activities/persistence/SocialTimelineCachePersistenceTest.java]: No license - and, as it's a new file,it won't get picked up by the other PR...

**Comment by ederign** [comment]
@csadilek  ready to merge!

**Comment by manstis** [comment]
+1 to merge.

File: kie-uberfire-social-activities/kie-uberfire-social-activities-backend/src/main/java/org/kie/uberfire/social/activities/security/SocialSecurityConstraintsManager.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+@ApplicationScoped
+public class SocialSecurityConstraintsManager {
+
+
+    @Inject
+    Instance<SocialSecurityConstraint> socialSecurityConstraints;
```
Comment (porcelli) [inline kie-uberfire-social-activities/kie-uberfire-social-activities-backend/src/main/java/org/kie/uberfire/social/activities/security/SocialSecurityConstraintsManager.java]: why not private?

**Comment by ederign** [comment]
updated: https://github.com/droolsjbpm/kie-uberfire-extensions/pull/13

**Pull request closed without merging**